celery/app/trace.py

```diff
@@ -150,6 +150,10 @@ class TraceInfo(object):
     def __init__(self, state, retval=None):
         self.state = state
         self.retval = retval
+        
+    @staticmethod
+    def get_name(task, req):
```

Comment:
If you modify this method signature to receive the name directly you can use it in both occurrences:

```
def get_task_name(name, request):
    return getattr(request, 'shadow', name)
```

The use of `getattr` is also a simplification here :smile: !

Reply:
Thought about that earlier. But the problem is that `request` in `trace_task()` is a dictionary, while `req` in `TraceInfo` is an object (`Context` in `celery/celery/app/task.py`). Sure, we can write more tests in the common function `get_task_name()` for both data types, but whether that's the best approach depends on the coding preference. Just my 2 cents.